test/suite/test_hs22.py

```diff
@@ -173,3 +177,108 @@ def test_out_of_order_timestamp_update_newer_than_tombstone(self):
+
+        if self.value_format == '8t':
+            value1 = 97  # 'a'
+            value2 = 98  # 'b'
+            value3 = 99  # 'c'
+            value4 = 100 # 'd'
+        else:
+            value1 = 'a'
+            value2 = 'b'
+            value3 = 'c'
+            value4 = 'd'
+
+        self.session.begin_transaction()
+        self.session.timestamp_transaction('commit_timestamp=' + self.timestamp_str(1))
+        cursor[key1] = value1
+        self.session.timestamp_transaction('commit_timestamp=' + self.timestamp_str(3))
+        cursor[key1] = value2
+        self.session.timestamp_transaction('commit_timestamp=' + self.timestamp_str(4))
+        cursor[key1] = value3
+        self.session.timestamp_transaction('commit_timestamp=' + self.timestamp_str(3))
+        cursor[key1] = value4
+        self.session.commit_transaction()
+
+        # Do a checkpoint to trigger
+        # history store reconciliation.
+        self.session.checkpoint()
+
+        evict_cursor = self.session.open_cursor(uri, None, "debug=(release_evict)")
+
+        # Search the key to evict it.
+        self.session.begin_transaction("read_timestamp=" + self.timestamp_str(10))
+        self.assertEqual(evict_cursor[key1], value4)
+        self.assertEqual(evict_cursor.reset(), 0)
+        self.session.rollback_transaction()
+
+        # Search the key again to verify the data is still as expected.
+        self.session.begin_transaction("read_timestamp=" + self.timestamp_str(10))
+        self.assertEqual(cursor[key1], value4)
+        self.session.rollback_transaction()
+
+    def test_out_of_order_timestamp_squash_updates(self):
+        uri = 'table:test_hs22'
+        format = 'key_format={},value_format={}'.format(self.key_format, self.value_format)
+        self.session.create(uri, format)
+        cursor = self.session.open_cursor(uri)
+        self.conn.set_timestamp(
+            'oldest_timestamp=' + self.timestamp_str(1) + ',stable_timestamp=' + self.timestamp_str(1))
+
+        key1 = self.key1
+
+        if self.value_format == '8t':
+            value1 = 97  # 'a'
+            value2 = 98  # 'b'
+            value3 = 99  # 'c'
+            value4 = 100 # 'd'
+            value5 = 101 # 'e'
+        else:
+            value1 = 'a'
+            value2 = 'b'
+            value3 = 'c'
+            value4 = 'd'
+            value5 = 'e'
+
+        self.session.begin_transaction()
+        cursor[key1] = value1
+        self.session.commit_transaction('commit_timestamp=' + self.timestamp_str(15))
+
+        self.session.begin_transaction()
+        cursor[key1] = value2
+        self.session.commit_transaction('commit_timestamp=' + self.timestamp_str(18))
+
+        self.session.begin_transaction()
+        cursor.set_key(key1)
+        cursor.remove()
+        cursor[key1] = value3
+        cursor[key1] = value4
+        cursor[key1] = value5
+        self.session.commit_transaction('commit_timestamp=' + self.timestamp_str(11))
```

Comment:
@Chenhao-Huang I believe that comparing the `prev_upd` and the `out_of_order_ts_upd` pointers to pop the OOO update from the OOO updates vector in `__wt_hs_insert_updates` is the correct approach. 

The reason is that in the case of multiple OOO updates(with the same txn and timestamp) the oldest one is chosen as an OOO update. Hence, we never miss the OOO update. Let me know if I am missing a specific scenario here.

Reply:
What about this case T @ 10 -> U @10 -> U@10 -> U@20?

Reply:
The thing worries me in your approach is that we have an implicit assumption that out_of_order_ts_upd should be the last upd in a series of upd from the same transaction id and timestamp. That can easily change in the future.

Reply:
In the case that you have mentioned, the oldest update with timestamp 10 will be the OOO update. This will be popped when `prev_upd` points to it.

Reply:
Am I correct in assuming that you are only referring to one case where the most recent OOO updates need squashing as a result of multiple updates having the same transaction and timestamp?
If I am correct in assuming the above then the worst that could happen if we choose the latest update with timestamp 10 is that we will be left with a OOO update in the stack.

Reply:
That is problematic. If we have more updates, like U@20 -> T@10 -> U@10 -> U@15. If 10 is not popped from the stack, we will fix U@20 with timestamp 10 as well, which is wrong.

Reply:
In the case you have mentioned above if let's say the latest update at timestamp 10 is OOO then it will be popped.

Reply:
Let try to add an assert to ensure the out of order ts pushed is the oldest of that timestamp or add a comment to describe this depencency.

Reply:
Updated in [3e8c92c](https://github.com/wiredtiger/wiredtiger/pull/7453/commits/3e8c92cb31b82a20d274cf10e8842bf00a29e9da)

Reply:
I suggest we run a mongodb patch build with this change since it is high risk.

Reply:
@quchenhao To ensure that this change works, I ran a WiredTiger and a MongoDB patch.
Note: The failing tests are also failing on the base commit and are not caused due to this change.

Reply:
@quchenhao I ran another [MongoDB patch](https://spruce.mongodb.com/version/61ee151a0ae606612bf03be2/tasks). Although there are a lot of failures, none of them is related to my change. Feel free to have a look.